File: frontend/src/types/reconciliation.ts
export interface VarianceRecord {
  id: string;
  reconciliation_run_id: string;
  node: string;
  expected_volume_m3: number;
  actual_volume_m3: number;
  variance_m3: number;
  variance_pct: number;
  is_exception: boolean;
  asset_id: string | null;
  reference_id: string | null;
  notes: string | null;
  fraud_checks: Record<string, unknown> | null;
  created_at: string;
}

export interface ReconciliationRun {
  id: string;
  name: string;
  status: string;
  run_type: string;
  period_start: string;
  period_end: string;
  asset_id: string | null;
  total_expected_m3: number | null;
  total_actual_m3: number | null;
  total_variance_pct: number | null;
  tolerance_threshold_pct: number;
  summary: string | null;
  completed_at: string | null;
  created_at: string;
  variance_records: VarianceRecord[];
}

File: backend/app/schemas/reconciliation.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.core.constants import ReconciliationNode, ReconciliationStatus


class ReconciliationRunCreate(BaseModel):
    name: str = Field(max_length=255)
    run_type: str = "MANUAL"
    period_start: datetime
    period_end: datetime
    asset_id: uuid.UUID | None = None
    tolerance_threshold_pct: float = 1.5


class ReconciliationRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, frozen=True)

    id: uuid.UUID
    name: str
    status: ReconciliationStatus
    run_type: str
    period_start: datetime
    period_end: datetime
    asset_id: uuid.UUID | None
    triggered_by_id: uuid.UUID | None
    total_expected_m3: float | None
    total_actual_m3: float | None
    total_variance_pct: float | None
    tolerance_threshold_pct: float
    summary: str | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    variance_records: list["VarianceRecordResponse"] = []


class VarianceRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, frozen=True)

    id: uuid.UUID
    reconciliation_run_id: uuid.UUID
    node: ReconciliationNode
    expected_volume_m3: float
    actual_volume_m3: float
    variance_m3: float
    variance_pct: float
    is_exception: bool
    asset_id: uuid.UUID | None
    reference_id: str | None
    notes: str | None
    fraud_checks: dict | None
    created_at: datetime


class ReconciliationTriggerRequest(BaseModel):
    name: str
    period_start: datetime
    period_end: datetime
    asset_id: uuid.UUID | None = None
    tolerance_threshold_pct: float = 1.5
